js/contexts/socketContext.tsx

```diff
@@ -0,0 +1,66 @@
+import { reload } from "../browser";
+import { join } from "../util/channel";
+import { getMetaContent } from "../util/metadata";
+import { Socket } from "phoenix";
+import {
+  createContext,
+  ReactNode,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { z } from "zod";
+
+const SocketContext = createContext<Socket | null>(null);
+
+const MetadataJoinResponse = z.object({
+  authenticated: z.boolean(),
+  server_release: z.string(),
+});
+
+export const SocketProvider = ({ children }: { children?: ReactNode }) => {
+  const [readySocket, setReadySocket] = useState<Socket | null>(null);
+
+  useEffect(() => {
+    const clientRelease = getMetaContent("release");
+    const initialSocket = new Socket("/socket", {
+      params: {
+        release: clientRelease,
+        token: getMetaContent("guardianToken"),
+      },
+    });
+    initialSocket.connect();
+
+    initialSocket.onError((err) => {
+      console.error(err);
+    });
+    const metadataChannel = initialSocket.channel("metadata");
```

Comment:
Connecting the socket is asynchronous, so the socket might not be connected yet. Will this automatically wait until the socket is connected before making and joining the channel? Or do you have to put it in a `socket.onOpen()` callback?

Reply:
Ooh, good question. In looking at the phoenix(js) source, channel creation doesn't cause any message send. And Channel#join, which uses Socket#push, is [buffered](https://github.com/phoenixframework/phoenix/blob/37c6085414daefdc88a4a303c876a616d425a4d7/assets/js/phoenix/socket.js#L607)!